ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -43,30 +45,39 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static com.github.ambry.cloud.CloudBlobMetadata.EncryptionOrigin;
 
 /**
  * The blob store that reflects data in a cloud storage.
  */
 class CloudBlobStore implements Store {
 
-  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private static final Logger logger = LoggerFactory.getLogger(CloudBlobStore.class);
   private final PartitionId partitionId;
   private final CloudDestination cloudDestination;
   private final CloudConfig cloudConfig;
+  private final CloudBlobCryptoAgent cryptoAgent;
+  private final VcrMetrics vcrMetrics;
   private final long minTtlMillis;
   private boolean started;
 
   /**
    * Constructor for CloudBlobStore
    * @param partitionId partition associated with BlobStore.
+   * @param cloudDestination the {@link CloudDestination}.
    * @param cloudConfig the {@link CloudConfig} to use.
    * @param cloudDestination the {@link CloudDestination} to use.
+   * @param cryptoAgent the {@link CloudBlobCryptoAgent} to use for encryption.
+   * @param vcrMetrics the {@link VcrMetrics} to use.
    */
-  CloudBlobStore(PartitionId partitionId, CloudConfig cloudConfig, CloudDestination cloudDestination) {
+  CloudBlobStore(PartitionId partitionId, CloudConfig cloudConfig, CloudDestination cloudDestination,
+      CloudBlobCryptoAgent cryptoAgent, VcrMetrics vcrMetrics) {
+    this.cloudDestination = Objects.requireNonNull(cloudDestination, "cloudDestination is required");
     this.partitionId = Objects.requireNonNull(partitionId, "partitionId is required");
     this.cloudConfig = Objects.requireNonNull(cloudConfig, "cloudConfig is required");
-    this.cloudDestination = Objects.requireNonNull(cloudDestination, "cloudDestination is required");
+    this.vcrMetrics = Objects.requireNonNull(vcrMetrics, "vcrMetrics is required");
     minTtlMillis = TimeUnit.DAYS.toMillis(cloudConfig.vcrMinTtlDays);
+    this.cryptoAgent = cryptoAgent;
```

Comment:
I feel it would be cleaner if we can generate cyrptoAgent based on config instead of from arguments. 

Looks like not straight forward but doable.

Reply:
I don't follow you.  The VcrServer instantiates CloudBlobCryptoAgentFactory from config properties and passes it to CloudBlobManager/CloudBlobStore.

Reply:
I see.  Yes, changed it to pass the agent factory to CloudBlobStore.

Reply:
Even factory, we can generate it inside `CloudBlobStore`, via `Utils.getObj()`.